[SmartAudio] Added support for custom frequency tables

functionpointer opened the pull request:
The idea of vtxTables is to allow the users to define their own frequency table that matches their local laws and conventions. This PR implements that.

In order to make custom frequency tables work, betaflight has to run the vtx in user frequency mode. This overrides the vtx's internal (fixed) frequency table. The downside of this is that it breaks channel display mechanisms built into the vtx, such as LEDs, built-in OSDs or CRSF telemetry.

Given that many users will just use the default frequency table, this is quite a limitation. This PR removes that limitation by verifying the contents of the frequency table. If vtxTable is set-up with a default frequency table, the vtx is run in band and channel mode. This results in a best of both worlds experience, offering either customizability or working built-in channel displays, depending on what the user wants.

To save flash space, the frequency table is verified using the CRC algorithm that is normally used for error detection in SmartAudio packets.

Please note that i am open to discussion. For example, i was thinking about exposing the mode selection (frequency or index mode) to the user, either in the CLI or the CMS. I opted for the CRC-based solution for its convenience and elegant simplicity.

mikeller commented on src/main/io/vtx_smartaudio.c:
```diff
@@ -762,6 +739,14 @@ bool vtxSmartAudioInit(void)
 
+    //This little trick saves us from storing the entire default frequency table just to find out if the frequencies are default or not.
+    //The default frequency table has a crc of 0x40. So if the current frequency table matches that, they are default.
+    uint8_t freqTableCRC = CRC8((uint8_t *)vtxSmartAudio.frequencyTable, sizeof(uint16_t) * vtxSmartAudio.capability.bandCount * vtxSmartAudio.capability.channelCount);
```
While I like this approach for its simplicity, I am not sure it is the right one - one of the main intentions behind VTX tables is to allow users to define custom frequency tables that restrict them to the frequencies that they are legally allowed to use in the jurisdiction they are flying in. In most places around the globe this will be a subset of the frequencies supported by the VTX hardware. If we do not want to encourage users to instead use the full frequency list supported by their VTX (and in the process use Betaflight to break the law), we should also support channel mode if the channels defined in the VTX table are all available as discrete channels in the VTX hardware - even if this means storing a full list - I don't think there is a cheaper way to achieve what we want.

functionpointer replied:
Well if the user wants to use the full frequency table he will still have to copy all frequencies into vtxTable. The CRC check just finds out if all the entered frequencies match the default table or not. An empty vtxTable will still prevent the vtx from working at all.

mikeller replied:
Yes - but pretty much no user will be able to use the full frequency table (that is in your proposal the only way to get channel mode to work) in a legal way, so adding this as it is will only serve to entice users to break the law, and should not be done.

functionpointer replied:
Ah you want a way to enable channel mode but with fewer frequencies. Sounds like a good idea.

We just have to deal with holes in the table properly. Lets take raceband for example. Suppose the local laws only allow the center four channels. In frequency mode, vtxTable could be set up like this:
`vtxtable band 5 RACEBAND R 5732 5769 5806 5843`
However, in index mode that would cause the vtx to go into channels 1-4, not 2-6 like we wanted.
So maybe this could be solved by introducing an empty placeholder like 0 or -1.
`vtxtable band 5 RACEBAND R 0 0 5732 5769 5806 5843 0 0`

mikeller replied:
> Ah you want a way to enable channel mode but with fewer frequencies.

Yep. Sorry for not making this clear the first time around.

The approach with a placeholder is definitely a way to get there. I am not sure that we even have to enforce that frequencies are in the 'correct' slot in a band in the VTX table - I see allowing users to create one or more 'mix and match' bands of all of the frequencies they want to use / are allowed to use in the VTX table as a feature. We could either do this by verifying that all of the frequencies are mappable on init, or even in a way that tried to map, and if possible used a discrete channel if available, or direct frequency entry otherwise, every time a new frequency was selected.
Both of these have the drawback that in the few cases where the same frequency is defined by more than one band it will not be possible to distinguish the bands - but I suspect this problem will occur for pretty much any approach that is less strict than the currently proposed 'full match' - even if we only allow empty placeholders (including entire bands) in the VTX table, it will not always be possible to match a VTX table band to a predefined band.

functionpointer replied:
Maybe we could have the user explicity define the mapping:
`vtxtable band 5 RACEBAND R R1(5658) R2(5695) 5555 F6(5840) A3(5825) 5648 R6(5843)`
This would eliminate the need for the default frequency table. I would have to figure out a way to store both channel and frequency information though.

mikeller replied:
This is getting quite complicated now.

What do you think about introducing a per band flag that allows the user to specify that this band is to be used as a 'factory band' and the frequency selected by sending the corresponding band / channel index to the VTX, with the frequencies for this band in the table becoming purely informational?

@jflyper, your thoughts?

functionpointer replied:
That sounds good, as long as it is in combination with empty placeholders. That way one could use index mode while still blocking illegal/unwanted frequencies. At the same time, unwanted factory bands could be overwritten with custom ones that use frequency mode.

The flexibility that my last suggestion would have provided was probably also overkill. After all, who wants a resorted raceband? Especially for interoperability it makes much more sense to just block unwanted frequencies but keep their positions the same.
In addition to the frequencies being in their stock positions, this would also require the bands to be in their stock positions. A small price to pay for the simplicity of this solution.

mikeller replied:
Good point about empty placeholders - setting a frequency to 0 should work just right for this.

Re stock positions - not really, the user just needs to know what bands channels are where, and fill in the right (indicative) frequencies, or else they will end up with a display that does not match the actual frequency.

functionpointer replied:
Exactly. So users will neither be able to change the order of factory bands nor their frequencies. Custom bands will have to squeeze into otherwise unwanted spots or at the end. But as i said, small price to pay.

That leaves one question though: What about other vtx protocols? IRC Tramp does not seem to have a band/channel mode. The easiest solution would be to declare factory bands as invalid for tramp devices. After all, when the frequency is only informational, no real freq can be given to the vtx.
Or should the factory flags be part of the SmartAudio implementation and thus by design separate from vtxtable?

mikeller replied:
I guess we'll need to at least support managing the 'factory band' flag at the higher VTX table levels. Protocols / drivers that do not support the concept of factory bands can then just completely skip doing anything when a user configures one and then tries to use it - part of the UX introduced by VTX tables is that it expects users to know what their hardware supports, and making it easy for them to do this by supporting the usage of pre-canned 'band plans'.

mikeller commented on src/main/io/vtx_smartaudio.c:
```diff
@@ -888,11 +873,47 @@ static bool vtxSAIsReady(const vtxDevice_t *vtxDevice)
+            saDevice.vtxTableFreqModeOn = true;
+            //device either currently in vtxtable freq mode, or in user freq mode but about to enter vtxtable freq mode
+            uint16_t freq = vtxCommonLookupFrequency(&vtxSmartAudio, band, channel);
+            if(saValidateFreq(freq)){
```
That old thing again - please use spaces after `if` and on the outside of curly braces.

functionpointer commented:
Alright i just pushed new changes. I have implemented the factory flag and it is being respected by all vtx implementations. Additionally, i have cleaned up the entire vtx control implementation.
In particular:
- moved vtxtable frequency mode implementation into `vtx_common.c`. This makes the implementation available for all vtx types and allows for some code deduplication (see point below)
I have contemplated doing this in `vtx.c` but i have decided against it as there is some other code that calls `vtxCommonGetBandAndChannel` and that code would have to call new functions in `vtx.c` instead. This would upset the `driver` and `io` structure somewhat.
- Removed band and channel tracking from tramp and rtc6705. The hardware underlying both only support frequency mode and the tracking is now done in `vtx_common.c` using the new factory flag.
- deleted vtxStringXXX. to continue supporting builds without `USE_VTX_TABLE`, new infrastructure was created in `drivers/vtx_table.c`, which loads fixed tables into vtxTableXXX when built witout `USE_VTX_TABLE`. Individual vtx implementations no longer need to load any band/channel tables. They only need to load their individual power tables when built without `USE_VTX_TABLE`. Additionally this allows for the next point:
- fully integrated vtxTableXXX and removed the old and no longer needed indirection of frequency and power tables in `vtxDevice_t`.
- removed VTX_SETTINGS_* constants from `vtx_common.h` and replaced them with the vtxtable equivalent.
- rtc6705 implementation now uses power values from vtxtable instead of using indices directly. It also stops using index 0. This makes it consistent with other vtx implementations and is more user configurable. It also cleans up `telemetry\srxl.c` which had to have a special case for rtc6705.

Let me know what you think!

RipperDrone commented:
@functionpointer Does it make sense to define a CLI variable allowing the user to switch vtxtable functionality ON or OFF instead of making it a pre-compile #define? Not sure myself, just found so many users confused not following 'RTFM' when vtxtable is switched on by default...

functionpointer commented:
No. @mikeller stated above that vtxtable should not entice users to break their local laws. A liberal default would be just that.

This whole PR exists to allow users to take advantage of all features of their hardware while only defining frequencies and power levels they are allowed to use. That way, nobody should have any reason to run with a config that allows them to break their local laws and regulations.

The pre-compilation option is just meant for flash space limited targets such as F3 that do not have enough space for vtxtable. `USE_VTX_TABLE` should be enabled whenever possible.

The RTFM problem should be dealt with some other way. Documentation in the wiki would be a good start. Another good step would be a tab in the betaflight configurator. Maybe some reminders could be placed in various locations. The cms menus for tramp and smartaudio already contain some.

mikeller commented:
@RipperDrone: I think the comment that 'so many users of unstable and untested development versions of the software are confused because they can not be bothered to read documentation' is pretty moot - if users want to use software that is easy to use and comes with good documentation then they should not be using unstable and untested development versions in the first place.

mikeller commented on src/main/cli/cli.c:
```diff
@@ -2817,8 +2815,13 @@ static void cliVtxTable(char *cmdline)
+        bool isFactory = false;
 
         for (channel = 0; channel <  channels && (tok  = strtok_r(NULL, " ", &saveptr)); channel++) {
+            if (channel == 0 && !isdigit(tok[0])) {
```
We can as well do a proper string comparison here, and throw an error if there is no match for `[FACTORY|USER]`.

functionpointer replied:
Makes sense, ill change it.
I have actually been thinking about the choice of `USER`. Maybe `CUSTOM` would fit better? Not that it makes much of a difference either way.

mikeller replied:
Yeah, `CUSTOM` makes more sense.

mikeller commented on src/main/io/vtx_rtc6705.c:
```diff
@@ -54,21 +54,26 @@ static vtxDevice_t vtxRTC6705 = {
+#ifndef USE_VTX_TABLE
+    //without USE_VTX_TABLE, fill vtxTable variables with default settings (instead of loading them from PG)
+    vtxTablePowerLevels = VTX_RTC6705_POWER_COUNT;
+    for(int i = 0; i < VTX_RTC6705_POWER_COUNT + 1; i++) {
```
Space after `for` missing.

mikeller commented on src/main/io/vtx_smartaudio.c:
```diff
@@ -825,13 +783,21 @@ static void vtxSAProcess(vtxDevice_t *vtxDevice, timeUs_t currentTimeUs)
+                    tfp_sprintf(saSupportedPowerLabels[i + 1], "%3d", constrain(saSupportedPowerValues[i], 0, 999));
                 }
             }
+            for(int8_t i = 0; i < vtxTablePowerLevels; i++) {
```
Space after `for` missing.

mikeller commented on src/main/io/vtx_smartaudio.c:
```diff
@@ -967,9 +959,12 @@ static void vtxSASetPitMode(vtxDevice_t *vtxDevice, uint8_t onoff)
+
     if (saValidateFreq(freq)) {
-        saSetMode(0);        //need to be in FREE mode to set freq
+        //saSetMode(0);        //need to be in FREE mode to set freq
```
If it's not needed then it should be removed.

mikeller commented on src/main/io/vtx_tramp.c:
```diff
@@ -620,29 +581,19 @@ bool vtxTrampInit(void)
+#ifndef USE_VTX_TABLE
+    //without USE_VTX_TABLE, fill vtxTable variables with default settings (instead of loading them from PG)
+    vtxTablePowerLevels = VTX_TRAMP_POWER_COUNT;
+    for(int i = 0; i < VTX_TRAMP_POWER_COUNT + 1; i++) {
```
Space after `for` missing.

functionpointer replied:
Alright i grew tired of it and managed to get astyle to run. Hopefully that will help me catch these things before pushing in the future. Thanks for your patience once again.

functionpointer commented:
Just fixed the formatting and rewrote the parser in commit https://github.com/betaflight/betaflight/pull/8380/commits/c20389d3f924f6872d939d396fc0822c796f96e0, and then rebased onto master with commit https://github.com/betaflight/betaflight/pull/8380/commits/854adf1cd5c2e94aff6c111aff4c3b559786a29f.

Looks like this time the old commits stay because they have the pull request id in their titles. Very cool.

functionpointer commented:
Excuse the mess, i made a mistake when squashing commits.
I fixed the formatting and rewrote the parser in commit https://github.com/betaflight/betaflight/commit/8c5eb8ea54d63b3dad94879da5cbfe3399fbe1c5, then rebased and squashed in commit https://github.com/betaflight/betaflight/commit/7cb34205b34642d3dddf2ac94219099f4e21c0a5.
